### PR TITLE
Staging to master to merge LigthGBM algo on Criteo

### PR DESCRIPTION
**Pull request by miguelgfierro**
### Description

Merging to master to add the work that @motefly did

### Related Issues

https://github.com/Microsoft/Recommenders/pull/633

### Checklist:

- [ ] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ The table below lists recommender algorithms available in the repository at the
 | [Smart Adaptive Recommendations (SAR)<sup>*</sup>](notebooks/00_quick_start/sar_movielens.ipynb) | Python CPU | Collaborative Filtering | Similarity-based algorithm for implicit feedback dataset |
 | [Surprise/Singular Value Decomposition (SVD)](notebooks/02_model/surprise_svd_deep_dive.ipynb) | Python CPU | Collaborative Filtering | Matrix factorization algorithm for predicting explicit rating feedback in datasets that are not very large | 
 | [Vowpal Wabbit Family (VW)<sup>*</sup>](notebooks/02_model/vowpal_wabbit_deep_dive.ipynb) | Python CPU (train online) | Collaborative, Content-Based Filtering | Fast online learning algorithms, great for scenarios where user features / context are constantly changing |
+| [LightGBM/Gradient Boosting Tree<sup>*</sup>](notebooks/00_quick_start/lightgbm_tinycriteo.ipynb) | Python CPU | Content-Based Filtering | Gradient Boosting Tree algorithm for fast training and low memory usage in content-based problems |
```
Comment (loomlike) [inline README.md]: Can we order the table alphabetically? I created an issue #663

Comment (miguelgfierro) [reply README.md]: We are now sorting as CPU->GPU->Spark. Do you mean to order alphabetically the algos within each env?

Comment (loomlike) [reply README.md]: Hmmmm... Let’s say one comes to find an algo for his problem:
1. Maybe already knows what he wants to use (algo name) but just looking for an implementation example of that algo. Then alphabetical sort will work for him.
2. Have no idea which algo to use, but knows about his problem which doesn’t have much historical data but bunch item features. Then algo types will be important to him and will skip CF-algos.

But I don’t see much benefit of sorting by CPU/GPU/spark, especially you can use CPU for GPU algos and vice versa anyway. Also you can use Databricks (spark) for CPU and GPU algos.

Comment (yueguoguo) [reply README.md]: Maybe it would be good to sort by `type` which is informative to users regarding different use cases.

Comment (loomlike) [reply README.md]: Github please give us interactive table...

Comment (loomlike) [reply README.md]: Github please give us interactive table...

**Comment by miguelgfierro** [comment]
Nice and clean error log from our dearest friend Java:
```
tests/unit/test_notebooks_pyspark.py ..F..                               [100%]

=================================== FAILURES ===================================
___________________________ test_als_deep_dive_runs ____________________________

notebooks = {'als_deep_dive': '/data/home/recocat/cicd/12/s/notebooks/02_model/als_deep_dive.ipynb', 'als_pyspark': '/data/home/re...aseline_deep_dive.ipynb', 'data_split': '/data/home/recocat/cicd/12/s/notebooks/01_prepare_data/data_split.ipynb', ...}

    @pytest.mark.notebooks
    @pytest.mark.spark
    def test_als_deep_dive_runs(notebooks):
        notebook_path = notebooks["als_deep_dive"]
>       pm.execute_notebook(notebook_path, OUTPUT_NOTEBOOK, kernel_name=KERNEL_NAME)

tests/unit/test_notebooks_pyspark.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/anaconda/envs/reco_pyspark/lib/python3.6/site-packages/papermill/execute.py:94: in execute_notebook
    raise_for_execution_errors(nb, output_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

nb = {'cells': [{'cell_type': 'code', 'metadata': {'inputHidden': True, 'hide_input': True}, 'execution_count': None, 'sour...nd_time': '2019-03-20T13:22:31.730071', 'duration': 34.491474, 'exception': True}}, 'nbformat': 4, 'nbformat_minor': 2}
output_path = 'output.ipynb'

    def raise_for_execution_errors(nb, output_path):
        """Assigned parameters into the appropriate place in the input notebook
    
        Parameters
        ----------
        nb : NotebookNode
           Executable notebook object
        output_path : str
           Path to write executed notebook
        """
        error = None
        for cell in nb.cells:
            if cell.get("outputs") is None:
                continue
    
            for output in cell.outputs:
                if output.output_type == "error":
                    error = PapermillExecutionError(
                        exec_count=cell.execution_count,
                        source=cell.source,
                        ename=output.ename,
                        evalue=output.evalue,
                        traceback=output.traceback,
                    )
                    break
    
        if error:
            # Write notebook back out with the Error Message at the top of the Notebook.
            error_msg = ERROR_MESSAGE_TEMPLATE % str(error.exec_count)
            error_msg_cell = nbformat.v4.new_code_cell(
                source="%%html\n" + error_msg,
                outputs=[
                    nbformat.v4.new_output(output_type="display_data", data={"text/html": error_msg})
                ],
                metadata={"inputHidden": True, "hide_input": True},
            )
            nb.cells = [error_msg_cell] + nb.cells
            write_ipynb(nb, output_path)
>           raise error
E           papermill.exceptions.PapermillExecutionError: 
E           ---------------------------------------------------------------------------
E           Exception encountered at "In [11]":
E           ---------------------------------------------------------------------------
E           Py4JJavaError                             Traceback (most recent call last)
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/IPython/core/interactiveshell.py in run_code(self, code_obj, result, async_)
E              3266                 else:
E           -> 3267                     exec(code_obj, self.user_global_ns, self.user_ns)
E              3268             finally:
E           
E           <ipython-input-11-6e4ac0080a04> in <module>
E                10 print(
E           ---> 11     "RMSE score = {}".format(evaluations.rmse()),
E                12     "MAE score = {}".format(evaluations.mae()),
E           
E           /data/home/recocat/cicd/12/s/reco_utils/evaluation/spark_evaluation.py in rmse(self)
E               118         """
E           --> 119         return self.metrics.rootMeanSquaredError
E               120 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/pyspark/mllib/evaluation.py in rootMeanSquaredError(self)
E               149         """
E           --> 150         return self.call("rootMeanSquaredError")
E               151 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/pyspark/mllib/common.py in call(self, name, *a)
E               145         """Call method of java_model"""
E           --> 146         return callJavaFunc(self._sc, getattr(self._java_model, name), *a)
E               147 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/pyspark/mllib/common.py in callJavaFunc(sc, func, *args)
E               122     args = [_py2java(sc, a) for a in args]
E           --> 123     return _java2py(sc, func(*args))
E               124 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in __call__(self, *args)
E              1256         return_value = get_return_value(
E           -> 1257             answer, self.gateway_client, self.target_id, self.name)
E              1258 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/pyspark/sql/utils.py in deco(*a, **kw)
E                62         try:
E           ---> 63             return f(*a, **kw)
E                64         except py4j.protocol.Py4JJavaError as e:
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
E               327                     "An error occurred while calling {0}{1}{2}.\n".
E           --> 328                     format(target_id, ".", name), value)
E               329             else:
E           
E           <class 'str'>: (<class 'py4j.protocol.Py4JNetworkError'>, Py4JNetworkError('An error occurred while trying to connect to the Java server (127.0.0.1:41589)',))
E           
E           During handling of the above exception, another exception occurred:
E           
E           IndexError                                Traceback (most recent call last)
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in _get_connection(self)
E               928         try:
E           --> 929             connection = self.deque.pop()
E               930         except IndexError:
E           
E           IndexError: pop from an empty deque
E           
E           During handling of the above exception, another exception occurred:
E           
E           ConnectionRefusedError                    Traceback (most recent call last)
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in start(self)
E              1066         try:
E           -> 1067             self.socket.connect((self.address, self.port))
E              1068             self.stream = self.socket.makefile("rb")
E           
E           ConnectionRefusedError: [Errno 111] Connection refused
E           
E           During handling of the above exception, another exception occurred:
E           
E           Py4JNetworkError                          Traceback (most recent call last)
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/IPython/core/interactiveshell.py in run_code(self, code_obj, result, async_)
E              3282             if result is not None:
E              3283                 result.error_in_exec = sys.exc_info()[1]
E           -> 3284             self.showtraceback(running_compiled_code=True)
E              3285         else:
E              3286             outflag = False
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/IPython/core/interactiveshell.py in showtraceback(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)
E              2021                                             value, tb, tb_offset=tb_offset)
E              2022 
E           -> 2023                     self._showtraceback(etype, value, stb)
E              2024                     if self.call_pdb:
E              2025                         # drop into debugger
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/ipykernel/zmqshell.py in _showtraceback(self, etype, evalue, stb)
E               544             u'traceback' : stb,
E               545             u'ename' : unicode_type(etype.__name__),
E           --> 546             u'evalue' : py3compat.safe_unicode(evalue),
E               547         }
E               548 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/ipython_genutils/py3compat.py in safe_unicode(e)
E                63     """
E                64     try:
E           ---> 65         return unicode_type(e)
E                66     except UnicodeError:
E                67         pass
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/protocol.py in __str__(self)
E               466     def __str__(self):
E               467         gateway_client = self.java_exception._gateway_client
E           --> 468         answer = gateway_client.send_command(self.exception_cmd)
E               469         return_value = get_return_value(answer, gateway_client, None, None)
E               470         return "{0}: {1}".format(self.errmsg, return_value)
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in send_command(self, command, retry, binary)
E               981          if `binary` is `True`.
E               982         """
E           --> 983         connection = self._get_connection()
E               984         try:
E               985             response = connection.send_command(command)
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in _get_connection(self)
E               929             connection = self.deque.pop()
E               930         except IndexError:
E           --> 931             connection = self._create_connection()
E               932         return connection
E               933 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in _create_connection(self)
E               935         connection = GatewayConnection(
E               936             self.gateway_parameters, self.gateway_property)
E           --> 937         connection.start()
E               938         return connection
E               939 
E           
E           /anaconda/envs/reco_pyspark/lib/python3.6/site-packages/py4j/java_gateway.py in start(self)
E              1077                 "server ({0}:{1})".format(self.address, self.port)
E              1078             logger.exception(msg)
E           -> 1079             raise Py4JNetworkError(msg, e)
E              1080 
E              1081     def _authenticate_connection(self):
E           
E           Py4JNetworkError: An error occurred while trying to connect to the Java server (127.0.0.1:41589)
```
it looks it is not finding localhost when calling RMSE? o_o

**Comment by gramhagen** [comment]
this happens if the local spark instance dies, i would think this is more related to memory issues or long-running jobs than the specific rmse call

File: README.md (hunk 1 above)
Comment (nikhilrj) [reply README.md]: Sorting first by type then by env seems like a good approach.

File: SETUP.md [hunk 2]
```diff
@@ -139,24 +136,26 @@ SPARK_WORKER_OPTS="-Dspark.worker.cleanup.enabled=true, -Dspark.worker.cleanup.a
 
 ## Setup guide for Azure Databricks
 
-### Requirements of Azure Databricks
+### Requirements
 
-* Runtime version 4.3 (Apache Spark 2.3.1, Scala 2.11)
+* Databricks Runtime version 4.3 (Apache Spark 2.3.1, Scala 2.11) or greater
```
Comment (nikhilrj) [inline SETUP.md]: Not related to this PR but CosmosDB now supports Spark 2.4. We should update our reqs to Databricks 5.0 - see #669.

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# LightGBM: A Highly Efficient Gradient Boosting Decision Tree\n",
+    "This notebook will give you a quick example of how to train LightGBM model in a recommendation scenario. \n",
+    "LightGBM \\[1\\] is a gradient boosting framework that uses tree-based learning algorithms. It is designed to be distributed and efficient with the following advantages:\n",
+    "* Faster training speed and higher efficiency.\n",
```
Comment (nikhilrj) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: "Faster" / "lower" / "better" are used without comparing LightGBM to anything. Can we be more clear? IE "better accuracy and faster training speed than _____"?

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+    "\n",
+    "<i>Licensed under the MIT License.</i>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# LightGBM: A Highly Efficient Gradient Boosting Decision Tree\n",
+    "This notebook will give you a quick example of how to train LightGBM model in a recommendation scenario. \n",
```
Comment (nikhilrj) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Can we provide more details here? "This notebook will give you a quick example of how to train a LightGBM model to estimate click-through rates on an e-commerce advertisement. We will train a LightGBM based model on the Criteo dataset."

File: notebooks/00_quick_start/README.md [hunk 5]
```diff
@@ -10,6 +10,7 @@ data preparation, model building, and model evaluation by using the utility func
 | [als](als_movielens.ipynb) | MovieLens | PySpark | Utilizing ALS algorithm to predict movie ratings in a PySpark environment.
 | [dkn](dkn_synthetic.ipynb) | Synthetic Data | Python CPU, GPU | Utilizing the Deep Knowledge-Aware Network (DKN) [2] algorithm for news recommendations using information from a knowledge graph, in a Python+GPU (TensorFlow) environment.
 | [fastai](fastai_movielens.ipynb) | MovieLens | Python CPU, GPU | Utilizing FastAI recommender to predict movie ratings in a Python+GPU (PyTorch) environment.
+| [lightgbm](lightgbm_tinycriteo.ipynb) | Criteo | Python CPU | Utilizing LigtGBM Boosting Tree to predict whether or not a user has clicked on an e-commerce ad |
```
Comment (nikhilrj) [inline notebooks/00_quick_start/README.md]: Typo: "LightGBM"

File: SETUP.md [hunk 6]
```diff
@@ -220,24 +229,16 @@ import reco_utils
...
+This repository includes an end-to-end example notebook that uses Azure Datbaricks to estimate a recommendation model using Alternating Least Squares, writes pre-computed recommendations to Azure Cosmos DB, and then creates a real-time scoring service that retrieves the recommendations from Cosmos DB. In order to execute that [notebook](notebooks/05_operationalize/als_movie_o16n.ipynb), you must install the Recommenders repository as a library (as described above), **AND** you must also install some additional dependencies. With the *Quick install* method, you just need to pass an additional option to the [installation script](scripts/databricks_install.py).
```
Comment (anargyri) [inline SETUP.md]: "Datbaricks" 
"using Alternating Least Squares" -> "using matrix factorization with Alternating Least Squares"

File: notebooks/00_quick_start/fastai_movielens.ipynb [hunk 7]
```diff
@@ -549,7 +558,7 @@
     "                     test_df=training_removed,\n",
     "                     user_col=USER, \n",
     "                     item_col=ITEM, \n",
-    "                     prediction_col=PREDICTION, top_k=TOP_K)\n",
+    "                     prediction_col=PREDICTION)\n",
```
Comment (anargyri) [inline notebooks/00_quick_start/fastai_movielens.ipynb]: Why remove top_k=K? Shouldn't this be explicit?

File: notebooks/00_quick_start/fastai_movielens.ipynb [hunk 8]
```diff
@@ -696,94 +708,174 @@
...
       }
      },
      "metadata": {},
      "output_type": "display_data"
     },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/data/anaconda/envs/reco_gpu/lib/python3.6/site-packages/ipykernel_launcher.py:11: DeprecationWarning: Function record is deprecated and will be removed in verison 1.0.0 (current version 0.19.0). Please see `scrapbook.glue` (nteract-scrapbook) as a replacement for this functionality.\n",
```
Comment (anargyri) [inline notebooks/00_quick_start/fastai_movielens.ipynb]: A lot of deprecation warnings have appeared in this notebook for some reason.

File: notebooks/00_quick_start/lightgbm_tinycriteo.ipynb [hunk 9]
```diff
@@ -0,0 +1,817 @@
+{
```
Comment (anargyri) [inline notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: TREE_LEARINING_RATE

Comment (anargyri) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: How about maximum depth of trees? It is an important parameter, you could mention the default value.

Comment (anargyri) [reply notebooks/00_quick_start/lightgbm_tinycriteo.ipynb]: Could we use one-hot encoding for the categorical features or would it be too slow?

**Pull request merged**